crates/wdl-engine/src/digest.rs

```diff
@@ -0,0 +1,291 @@
+//! Implements calculation of file and directory content digests.
+//!
+//! This is used by the call cache and for uploading inputs for remote backends.
+
+use std::collections::HashMap;
+use std::path::Path;
+use std::sync::Arc;
+use std::sync::LazyLock;
+use std::sync::Mutex;
+
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::bail;
+use arrayvec::ArrayString;
+use blake3::Hash;
+use blake3::Hasher;
+use cloud_copy::ContentDigest;
+use cloud_copy::UrlExt;
+use tokio::sync::OnceCell;
+use tokio::task::spawn_blocking;
+use url::Url;
+use walkdir::WalkDir;
+
+use crate::ContentKind;
+use crate::cache::Hashable;
+use crate::http::Transferer;
+use crate::path::EvaluationPath;
+
+/// The variant tag for files.
+const FILE_VARIANT_TAG: u8 = 0;
+/// The variant tag for directories.
+const DIRECTORY_VARIANT_TAG: u8 = 1;
+
+/// Represents a calculated [Blake3](https://github.com/BLAKE3-team/BLAKE3) digest of a file or directory.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Digest {
+    /// The digest is for a file.
+    File(Hash),
+    /// The digest is for a directory.
+    Directory(Hash),
+}
+
+impl Digest {
+    /// Converts the digest to a hex string.
+    pub fn to_hex(self) -> ArrayString<64> {
+        match self {
+            Self::File(hash) => hash.to_hex(),
+            Self::Directory(hash) => hash.to_hex(),
+        }
+    }
+}
+
+/// Keeps track of previously calculated digests.
+///
+/// As WDL evaluation cannot write to existing files, it is assumed that files
+/// and directories are not modified during evaluation.
```

Comment:
Sadly this is not strictly true given the prevalence of read-write input mounts in other engines, but someone would really have to be trying to self-sabotage in order to witness the potential effects of this. But regardless, we have to trust them not to mess with files through some other external process.

Reply:
This comment was meant to only apply to Sprocket; we aren't really concerned with what other engines do in this context.

If a user happened to be running Sprocket at the same time as some other engine that _does_ allow WDL to write to existing files, firstly Sprocket would likely fail its evaluation (depending on how the user's command is authored) and secondly it would be no different than any other external process modifying the files during our evaluation. 

That said, it'll be true of other engines soon enough as that behavior is fully deprecated in 1.2 and required not to occur in a future 2.0.